### ISSUE TITLE
Add methods for mixture of word-vectors

### ISSUE THREAD

**Issue by menshikh-iv** [RaRe-Technologies/gensim#1879]
**Intro**
Users want to construct document vector from word-vectors (that already exist). As I see now, we have a really huge number of high-quality word-vectors (from Stanford, Google, Numberbatch, etc), for this reason, this feature can be really useful 

**To Do**
1. Collect ways, how we can mix-up word-vectors to document vector
    - the average of all word-vectors
    - the average with weights (like TfIdf or some custom weights from a user)
    - https://openreview.net/pdf?id=SyK00v5xx (IMO most powerful method here)
    - something else (investigate more methods is also part of this step)
2. Add this approaches as one (or several) methods to gensim (to "vector-storage classes") 
3. Write a Notebook with demonstration (how you can use it) + benchmarking (comparing this technique with document-vectors on several tasks, good benchmark reference is https://openreview.net/pdf?id=SyK00v5xx).

Based on https://groups.google.com/forum/#!topic/gensim/OfBjD0GU2xA (this is the latest thread about it)

**Comment by poornagurram** [comment]
I have done averaging word vectors. Can I take up this task?

**Comment by menshikh-iv** [comment]
Hello @poornagurram, @kakshay21 already work on it, but any help will be appreciated (discussion, proposing some methods, useful links, etc).

**Comment by poornagurram** [comment]
I would like to help. Let me connect and discuss with him.
I would like to think if we can, even more, enhance this and would help in implementing that.

**Comment by kakshay21** [comment]
Embeddings words and sentences via character n-grams
https://arxiv.org/abs/1607.02789
* outperforms PARAGRAM-PHRASE, RNNs and CNNs.
* CHARAGRAM models embeds a character sequence x by adding the vectors of its character n-grams followed by an element wise nonlinearity.
* this model converges  fastest to high performance.

**Comment by kakshay21** [comment]
Deep sentence embedding using long short-term memory networks 
https://arxiv.org/pdf/1502.06922.pdf
* It mainly embeds keywords in the final semantic vector representing the whole sentence.
* This paper seems to performed better in query sentences in search engines where it’s weakly supervised based on the clicks of the user.
* LSTM-RNN performs better than bare RNN for this task.
* Seems to outperform doc2vec and the result is similar to RNN with GRU.
* Lots of fancy word that I don't understand 😂

**Comment by kakshay21** [comment]
Unsupervised Learning of Sentence Embedding using Composition n-Gram Features
https://arxiv.org/pdf/1703.02507.pdf
* The sentence embedding is defined as the average of the source word embeddings of its constituent words 
* Also augmented the model by  learning source embeddings for not only unigrams but also n-grams present in each sentence, and averaging the n-gram embeddings along with the words
* seems to outperform C-PHRASE model (http://clic.cimec.unitn.it/marco/publications/acl2015/pham-et-al-cphrase-acl2015.pdf)
* achieves better performance than all other un- supervised competitors except the SkipThought vectors. 
Advantage
* low computational cost for both inference and training. 
* Sent2Vec models are also faster to train when compared to methods like SkipThought and DictRep owing to the SGD step allowing a high degree of paralleliz- ability.

**Comment by kakshay21** [comment]
@menshikh-iv  After referring to these papers, I see the trend of averaging the word vectors for sentence embedding is quite common.

**Comment by poornagurram** [comment]
Averaging is being used since long time and it's good to have. In one of the kaggle competition I saw a solution with averaging for quora question pairs with good amount of accuracy.

**Comment by kakshay21** [comment]
@poornagurram can you point me to that solution just for the reference?

**Comment by poornagurram** [comment]
Yeah sure that is in this [blog post](https://www.linkedin.com/pulse/duplicate-quora-question-abhishek-thakur/)

**Comment by kakshay21** [comment]
Adding to the previous finding, with specific details

1. Unsupervised Learning of Sentence Embeddings using Compositional n-Gram Features
link to the paper - https://arxiv.org/pdf/1703.02507.pdf
Algorithm - 
![screen shot 2018-02-20 at 4 01 36 pm](https://user-images.githubusercontent.com/13598994/36424586-c9d78fba-1669-11e8-98d3-3d4487a2d612.png)
Where R(S) is the list of n-grams including unigrams present in sentence S.
code 

2. Siamese CBOW: Optimizing Word Embeddings for Sentence Representations
link to the paper - http://aclweb.org/anthology/P/P16/P16-1089.pdf
 It takes in the average of word embeddings and predicts the sentence embedding.
Algorithm -
![screen shot 2018-02-20 at 4 11 12 pm](https://user-images.githubusercontent.com/13598994/36424620-f88dd6ac-1669-11e8-8d1e-6a1f876972e0.png)￼
Probability function 
![screen shot 2018-02-20 at 4 13 59 pm](https://user-images.githubusercontent.com/13598994/36424624-ff1ccbf4-1669-11e8-990a-b11b222f6d5d.png)
Loss function￼
![screen shot 2018-02-20 at 4 14 13 pm](https://user-images.githubusercontent.com/13598994/36424628-02fed8d4-166a-11e8-9347-c17d6aa98ec2.png)

3. A SIMPLE BUT TOUGH-TO-BEAT BASELINE FOR SENTENCE EMBEDDINGS
link to the paper - https://openreview.net/pdf?id=SyK00v5xx
Algorithm -
![screen shot 2018-02-20 at 6 23 19 pm](https://user-images.githubusercontent.com/13598994/36424968-30e37e70-166b-11e8-877b-026487c8c09c.png)

**Comment by gojomo** [comment]
A module related to working with sets-of-words, enhanced by word-vectors, could also absorb the things like "Word Mover's Distance" or "Soft Cosine Similarity" that currently bloat the `KeyedVectors` class a bit.
